### PR TITLE
fix: Emits connected event on a refresh event

### LINKED ISSUE #337: Unauthorized through CIDR restrictions on the secret ID
### Bug Description

See discussion [here](https://matrix.to/#/!yAkGlrYcBFYzYRvOlQ:ubuntu.com/$iSh81WI9kAhzAnXzDoPjm4_yGu61Qcb8p7GnoC8Zeo8?via=ubuntu.com&via=matrix.org&via=mozilla.org)
```

### To Reproduce

See discussion [here](https://matrix.to/#/!yAkGlrYcBFYzYRvOlQ:ubuntu.com/$iSh81WI9kAhzAnXzDoPjm4_yGu61Qcb8p7GnoC8Zeo8?via=ubuntu.com&via=matrix.org&via=mozilla.org)

### Environment

microk8s

### Relevant log output

```shell
https://pastebin.canonical.com/p/Xk9qjgfCWG/
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

The aim is to allow the requirer charm to provide a refresh event that will trigger the `connected` event where the requirer can re-request the credentials with new IPs.

Fixes #337 

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (gruyaume) [inline lib/charms/vault_k8s/v0/vault_kv.py:542]: I'm not sure I understand why we observe config changed as a default? Nothing here depends on the requiring charm config right? We may want to use `update status` instead.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py:566]: This change is initiated by a unit going down then coming up again with a new IP address.
Based on the [documentation](https://juju.is/docs/sdk/config-changed-event) of the `config-changed` event it seems to me that a unit changing its IP will trigger a `config-changed` event.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: This change is initiated by a unit going down then coming up again with a new IP address.
Based on the [documentation](https://juju.is/docs/sdk/config-changed-event) of the `config-changed` event it seems to me that a unit changing its IP will trigger a `config-changed` event.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py]: Agreed. However it is really difficult to come up with a good name here. I am convinced some more in depth refactoring is needed. But I'd keep it out of the scope of this bugfix.

File: tests/unit/lib/charms/vault_k8s/v0/test_vault_kv.py
Comment (Kayra1) [inline tests/unit/lib/charms/vault_k8s/v0/test_vault_kv.py:402]: We don't need the [-1] here, we're not assigning it to anything.

**Review by Kayra1 — APPROVED** [review]
Good job. One minor nit and I would wait for one more approval also.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (Kayra1) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: How do you imagine people would use `refresh_events`? Isn't it enough to just look at `config_changed` since that fires when the IP changes?

Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: Users don't need to use `refresh_event`, they can specify one depending on their use case. And if they don't we jut use the `config_changed` event.

Comment (Kayra1) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: I'm wondering if they will ever need or want to specify one. What use case is there for it? can you give an example?

File: tests/unit/lib/charms/vault_k8s/v0/test_vault_kv.py
Comment (saltiyazan) [reply tests/unit/lib/charms/vault_k8s/v0/test_vault_kv.py]: Copy paste :D Thanks

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: No, I don't have a use case on top of my mind.
We can silently do `config-changed` in the background if we are convinced that users won't change it.
But I don't mind keeping the flexibility for example if a user doesn't want to trigger this handler on each config change.
what do you think @Kayra1 ?

Comment (Kayra1) [reply lib/charms/vault_k8s/v0/vault_kv.py:542]: I think we should remove this option entirely and do it silently with just `config-changed`. It's one less thing for users to worry about and there doesn't seem to be a catch with it that users need to code around or know about.

We can introduce it in the future if someone needs it, but I can't imagine why anyone would want to create a custom event just for one of their libraries either.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (gruyaume) [inline lib/charms/vault_k8s/v0/vault_kv.py:536]: The fact that the default is None is somewhat misleading as if it will be set to `config_changed` in this case.  If we want it to be config changed, why not place it here and explain why it's the default in the docstring (which I realise we don't have at the moment)

**Review by gruyaume — COMMENTED** [review]
I agree with most of this change, it's just the handle_refresh that I think we can improve a bit.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (Kayra1) [reply lib/charms/vault_k8s/v0/vault_kv.py:536]: in my opinion this argument can be deleted and only config changed should be observed in the library, but if we really want it to stay I agree that it should default to config changed.

Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py:536]: @gruyaume Do you have any opinions on completely removing it as @Kayra1 suggests? I'd keep it at least for the flexibility it gives but on the other hand I don't have a use case in mind where users would chose a different event.

Comment (gruyaume) [reply lib/charms/vault_k8s/v0/vault_kv.py:536]: If we don't have a use case for it, let's not add this unneeded flexibility. And if we keep the config_changed observing, let's mention somewhere why we do it because it's not straightforward why we care about this event.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py]: Event removed.

File: lib/charms/vault_k8s/v0/vault_kv.py
Comment (saltiyazan) [reply lib/charms/vault_k8s/v0/vault_kv.py]: Done.

**Review by gruyaume — APPROVED** [review]

**Pull request merged**
